test/powershell/Modules/Microsoft.PowerShell.Utility/Add-Type.Tests.ps1

```diff
@@ -28,41 +28,242 @@ Describe "Add-Type" -Tags "CI" {
+
+        # Add-Type show parse and compile errors and then finish with an terminationg error.
+        # Catch non-termination information error.
+        { Add-Type -MemberDefinition $sourceCode -Name $sourceType -Namespace $sourceNS -Language $sourceLanguage -ErrorAction Stop } | Should -Throw -ErrorId "SOURCE_CODE_ERROR,Microsoft.PowerShell.Commands.AddTypeCommand"
+        # Catch final terminationg error.
+        { Add-Type -MemberDefinition $sourceCode -Name $sourceType -Namespace $sourceNS -Language $sourceLanguage -ErrorAction SilentlyContinue } | Should -Throw -ErrorId "COMPILER_ERRORS,Microsoft.PowerShell.Commands.AddTypeCommand"
+
+        $returnedTypes = Add-Type -MemberDefinition $sourceCode -Name $sourceType -UsingNamespace $sourceUsingNS -Namespace $sourceNS -Language $sourceLanguage -PassThru
+        ([type]$sourceRunType)::TestString() | Should BeExactly $expectedResult
 
-        { [Test.AddType.BasicTest1]::Add1(1, 2) } | Should -Throw -ErrorId "TypeNotFound"
-        { [Test.AddType.BasicTest2]::Add2(3, 4) } | Should -Throw -ErrorId "TypeNotFound"
+        # Return the same assembly if source code has not been changed.
+        $returnedTypes2 = Add-Type -MemberDefinition $sourceCode -Name $sourceType -UsingNamespace $sourceUsingNS -Namespace $sourceNS -Language $sourceLanguage -PassThru
+        $returnedTypes[0].Assembly.FullName | Should BeExactly $returnedTypes2[0].Assembly.FullName
+
+        # With default namespace.
+        Add-Type -MemberDefinition $sourceCode -Name $sourceType -UsingNamespace $sourceUsingNS -Language $sourceLanguage
+        ([type]$sourceDefaultNSRunType)::TestString() | Should BeExactly $expectedResult
+    }
 
-        Add-Type -Path $codeFile1,$codeFile2
+    It "Can compile without loading" {
+
+        ## The assembly files cannot be removed once they are loaded, unless the current PowerShell session exits.
+        ## If we use $TestDrive here, then Pester will try to remove them afterward and result in errors.
+        $TempPath = [System.IO.Path]::GetTempFileName()
+        if (Test-Path $TempPath) { Remove-Item -Path $TempPath -Force -Recurse }
+        New-Item -Path $TempPath -ItemType Directory -Force > $null
+
+        $outFile = Join-Path -Path $TempPath -ChildPath "assembly$guid.dll"
+        $outFile2 = Join-Path -Path $TempPath -ChildPath "assembly2$guid.dll"
+
+        $code = @"
+using System.Management.Automation;
+[System.Management.Automation.Cmdlet("Get", "CompileThing$guid", ConfirmImpact = System.Management.Automation.ConfirmImpact.High, SupportsPaging = true)]
+public class AttributeTest$guid : PSCmdlet
+{
+    protected override void EndProcessing()
+
+    {
+        WriteObject("$guid");
+    }
+}
+"@
+
+        $cmdlet = "Get-CompileThing$guid"
+
+        Add-Type -TypeDefinition $code -OutputAssembly $outFile | Should -BeNullOrEmpty
```

Comment:
Could you please add a test to verify that when `-OutputAssembly` is specified, the generated assembly is not loaded by default, unless `-PassThru` is specified too?

Reply:
1. Added
2. Please see Line 153.

Reply:
1. Added
2. Please see Line 153.

Reply:
1. Added
2. Please see Line 153.